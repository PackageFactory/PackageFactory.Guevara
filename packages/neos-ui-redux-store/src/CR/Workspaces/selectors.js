import {$get} from 'plow-js';
import {createSelector} from 'reselect';

export const personalWorkspaceNameSelector = $get('cr.workspaces.personalWorkspace.name');

export const activeDocumentContextPathSelector = $get('ui.contentCanvas.contextPath');

export const baseWorkspaceSelector = $get('cr.workspaces.personalWorkspace.baseWorkspace');

export const isWorkspaceReadOnlySelector = $get('cr.workspaces.personalWorkspace.readOnly');

export const publishableNodesSelector = $get('cr.workspaces.personalWorkspace.publishableNodes');

export const publishableNodesInDocumentSelector = createSelector(
    [
        publishableNodesSelector,
        activeDocumentContextPathSelector
    ],
    (publishableNodes, activeDocumentContextPath) => publishableNodes.filter(i => $get('documentContextPath', i) === activeDocumentContextPath)
);

export const makeIsDocumentNodeDirtySelector = () => createSelector(
    [
        publishableNodesSelector,
        (state, documentContextPath) => documentContextPath
    ],
    (publishableNodes, documentContextPath) => publishableNodes.filter(i => (
        $get('documentContextPath', i) === documentContextPath ||
        $get('contextPath', i) === documentContextPath
    )).count() > 0
);

export const makeIsContentNodeDirtySelector = () => createSelector(
    [
        publishableNodesSelector,
        (state, contextPath) => contextPath
    ],
    (publishableNodes, contextPath) => publishableNodes.filter(i => $get('contextPath', i) === contextPath).count() > 0
);
