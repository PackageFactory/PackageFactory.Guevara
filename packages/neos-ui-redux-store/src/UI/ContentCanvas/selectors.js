import {createSelector} from 'reselect';
import {$get} from 'plow-js';

export const shouldScrollIntoView = $get('ui.contentCanvas.shouldScrollIntoView');
export const getCurrentContentCanvasContextPath = $get('ui.contentCanvas.contextPath');
export const currentlyEditedPropertyName = $get('ui.contentCanvas.currentlyEditedPropertyName');

export const formattingUnderCursor = createSelector(
    [
        $get('ui.contentCanvas.formattingUnderCursor')
    ],
    formattingUnderCursor =>
        formattingUnderCursor.toJS()
);

export const documentNodeSelector = createSelector(
    [
        $get('ui.contentCanvas.contextPath'),
        $get('cr.nodes.byContextPath')
    ],
    (documentNodeContextPath, nodesByContextPath) => $get(documentNodeContextPath, nodesByContextPath)
);
