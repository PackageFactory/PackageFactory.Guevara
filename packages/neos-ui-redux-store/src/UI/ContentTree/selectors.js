import {$get} from 'plow-js';

export const getUncollapsed = $get('ui.contentTree.uncollapsed');
export const getIsLoading = $get('ui.contentTree.isLoading');
