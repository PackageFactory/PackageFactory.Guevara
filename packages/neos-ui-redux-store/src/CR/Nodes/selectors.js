import {$get} from 'plow-js';
import {createSelector, defaultMemoize} from 'reselect';
import {getCurrentContentCanvasContextPath} from './../../UI/ContentCanvas/selectors';

const nodes = $get(['cr', 'nodes', 'byContextPath']);
const focused = $get('cr.nodes.focused.contextPath');
const hovered = $get('cr.nodes.hovered.contextPath');

const parentNodeContextPath = contextPath => {
    if (typeof contextPath !== 'string') {
        return null;
    }

    const [path, context] = contextPath.split('@');

    return `${path.substr(0, path.lastIndexOf('/'))}@${context}`;
};

export const isDocumentNodeSelectedSelector = createSelector(
    [
        focused,
        getCurrentContentCanvasContextPath
    ],
    (focused, currentContentCanvasContextPath) => {
        return !focused || (focused === currentContentCanvasContextPath);
    }
);

// PERFORMANCE: This helper method is NOT allowed to post-process the retrieved node in any way;
// as we need to ensure the output is deterministic and can be cached for upstream selectors to
// work correctly.
//
// Instead of calling "node.toJS()" here (which totally breaks performance), we need to
// use the immutableNodeToJs() at the USAGE POINT of "nodeByContextPath", i.e.:
//
// nodeByContextPath (re-calculated at all times)
//    ---> focusedSelector (also re-calculated at all times, because nodeByContextPath changes for all invocations); but
//                         the RESULT of focusedSelector is "stable" again.
//    ---> immutableNodeToJs(focusedSelector): converts the focusedSelector result to plain JS; properly using memoization here.
const nodeByContextPath = state => contextPath =>
    $get(['cr', 'nodes', 'byContextPath', contextPath], state);

const immutableNodeToJs = selectorWhichEmitsNode =>
    createSelector(
        [
            selectorWhichEmitsNode
        ],
        node => {
            if (node && node.toJS) {
                node = node.toJS();
            }
            return node;
        }
    );

export const makeGetDocumentNodes = nodeTypesRegistry => createSelector(
    [
        nodes
    ],
    nodesMap => {
        const documentSubNodeTypes = nodeTypesRegistry.getSubTypesOf('TYPO3.Neos:Document');

        return nodesMap.filter(node => documentSubNodeTypes.includes(node.get('nodeType')));
    }
);

export const focusedNodePathSelector = immutableNodeToJs(createSelector(
    [
        focused,
        getCurrentContentCanvasContextPath
    ],
    (focused, currentContentCanvasContextPath) => {
        return focused || currentContentCanvasContextPath;
    }
));

export const focusedSelector = immutableNodeToJs(createSelector(
    [
        focusedNodePathSelector,
        nodeByContextPath
    ],
    (focusedNodePath, getNodeByContextPath) =>
        getNodeByContextPath(focusedNodePath)
));

export const hoveredSelector = immutableNodeToJs(createSelector(
    [
        hovered,
        nodeByContextPath
    ],
    (hoveredNodePath, getNodeByContextPath) =>
        getNodeByContextPath(hoveredNodePath)
));

export const byContextPathSelector = defaultMemoize(
    contextPath => immutableNodeToJs(createSelector(
        [
            nodeByContextPath
        ],
        getNodeByContextPath => getNodeByContextPath(contextPath)
    ))
);

export const parentNodeSelector = state => baseNode =>
    byContextPathSelector(parentNodeContextPath(baseNode.contextPath))(state);

export const grandParentNodeSelector = state => baseNode =>
    byContextPathSelector(parentNodeContextPath(parentNodeContextPath(baseNode.contextPath)))(state);
