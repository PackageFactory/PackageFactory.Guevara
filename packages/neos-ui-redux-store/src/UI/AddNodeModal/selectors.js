import {$get} from 'plow-js';
import {createSelector} from 'reselect';
import {selectors as nodes} from '../../CR/Nodes/index';

const referenceNodeContextPathSelector = state => $get('ui.addNodeModal.referenceNode', state);

export const referenceNodeSelector = createSelector(
    [
        referenceNodeContextPathSelector,
        state => state
    ],
    (referenceNodeContextPath, state) =>
        nodes.byContextPathSelector(referenceNodeContextPath)(state)
);

export const referenceNodeParentSelector = createSelector(
    [
        referenceNodeSelector,
        state => state
    ],
    (referenceNode, state) => {
        if (!referenceNode) {
            return undefined;
        }

        return nodes.parentNodeSelector(state)(referenceNode);
    }
);

export const referenceNodeGrandParentSelector = createSelector(
    [
        referenceNodeParentSelector,
        state => state
    ],
    (referenceNodeParent, state) => {
        if (!referenceNodeParent) {
            return undefined;
        }

        return nodes.parentNodeSelector(state)(referenceNodeParent);
    }
);
