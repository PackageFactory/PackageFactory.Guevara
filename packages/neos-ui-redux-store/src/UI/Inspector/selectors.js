import {map, mapObjIndexed, values, sort, compose} from 'ramda';
import {createSelector, defaultMemoize} from 'reselect';
import {$get, $count, $transform} from 'plow-js';
import validate from '@neos-project/neos-ui-validators/src/index';
import {selectors as nodes} from '../../CR/Nodes/index';

/**
 * Transient values and everything which depends on this
 *
 * - transientValues
 * - propertiesForValidation (merges transientValues + selectors.CR.Nodes.focusedSelector)
 * - makeValidationErrorsSelector (contains the full validation errors)
 */
export const transientValues = createSelector(
    [
        nodes.focusedNodePathSelector,
        $get('ui.inspector.valuesByNodePath')
    ],
    (focusedNodeContextPath, valuesByNodePath) => $get([focusedNodeContextPath], valuesByNodePath)
);

export const isDirty = createSelector(
    [
        transientValues
    ],
    transientValues => Boolean(transientValues && ($count('transientValues', {transientValues}) > 0))
);

export const shouldPromptToHandleUnappliedChanges = $get('ui.inspector.shouldPromptToHandleUnappliedChanges');

export const shouldShowSecondaryInspector = $get('ui.inspector.secondaryInspectorIsOpen');

const propertiesForValidationSelector = createSelector(
    [

        transientValues,
        nodes.focusedSelector
    ],
    (
        transientValues,
        focusedNode
    ) => {
        const propertiesForValidation = Object.assign({}, $get('properties', focusedNode));
        if (transientValues) {
            // Override values with transient values
            Object.keys(transientValues.toJS()).forEach(key => {
                propertiesForValidation[key] = $get([key, 'value'], transientValues);
            });
        }
        return propertiesForValidation;
    }
);

// here, we use "defaultMemoize()", as validation errors are quasi-singletons; so we want all instances of the selector
// to be shared for all fields. If we did not do this, the system would recompute the validation for each form field (which
// is way too often.)
export const makeValidationErrorsSelector = defaultMemoize((nodeTypesRegistry, validatorRegistry) => createSelector(
    [
        propertiesForValidationSelector,
        nodes.focusedSelector
    ],
    (
        propertiesForValidation,
        focusedNode
    ) => {
        if (!focusedNode) {
            return false;
        }

        const nodeType = nodeTypesRegistry.get($get('nodeType', focusedNode));
        if (!$get('properties', nodeType)) {
            console.error(`No properties configured in ${$get('nodeType', focusedNode)} nodetype`);
            return null;
        }

        const validationErrors = validate(propertiesForValidation, $get('properties', nodeType), validatorRegistry);

        return validationErrors;
    }
));

export const makeIsApplyDisabledSelector = (nodeTypesRegistry, validatorRegistry) => createSelector(
    [
        isDirty,
        makeValidationErrorsSelector(nodeTypesRegistry, validatorRegistry)
    ],
    (
        isDirty,
        validationErrors
    ) => {
        return !isDirty || (isDirty && validationErrors !== null);
    }
);

export const isDiscardDisabledSelector = createSelector(
    [
        isDirty
    ],
    isDirty => !isDirty
);

/*
    Will create a configuration for the Inspector Component to render itself
    with the following shape:

    {
        "tabs": [
            {
                "id": "my-tab-1",
                "icon": "icon-cog",
                "groups": [
                    {
                        "id": "my-group-1",
                        "label": "MyGroup 1",
                        "properties": {
                            "id": "my-property-1",
                            "label": "MyProperty 1",
                            "editor": "MyAwesome.Package:MyEditor"
                        }
                    }
                ]
            }
        ]
    }
*/
const toJS = val => val && val.toJS ? val.toJS() : val;
const withId = mapObjIndexed((subject, id) => ({
    ...subject,
    id
}));
const getPosition = subject => subject.ui ? subject.ui.position : subject.position;
const positionalArraySorter = sort((a, b) => (getPosition(a) - getPosition(b)) || (a.id - b.id));
const getNormalizedDeepStructureFromNode = path => compose(
    positionalArraySorter,
    values,
    withId,
    toJS,
    $get(path)
);

const getTabs = getNormalizedDeepStructureFromNode('nodeType.ui.inspector.tabs');
const getGroups = getNormalizedDeepStructureFromNode('nodeType.ui.inspector.groups');
const getProperties = getNormalizedDeepStructureFromNode('nodeType.properties');

export const viewConfiguration = createSelector(
    [
        nodes.focusedSelector
    ],
    node => {
        if (!node) {
            return undefined;
        }
        const tabs = getTabs(node);
        const groups = getGroups(node);
        const properties = getProperties(node);

        return {
            tabs: map(
                tab => ({
                    ...tab,
                    groups: map(
                        group => ({
                            ...group,
                            properties: map(
                                $transform({
                                    id: $get('id'),
                                    label: $get('ui.label'),
                                    editor: $get('ui.inspector.editor'),
                                    editorOptions: $get('ui.inspector.editorOptions')
                                }),
                                properties.filter(p => $get('ui.inspector.group', p) === group.id)
                            )
                        }),
                        groups.filter(g => {
                            const isMatch = g.tab === tab.id;
                            const isDefaultTab = !g.tab && tab.id === 'default';

                            return isMatch || isDefaultTab;
                        })
                    )
                }),
                tabs
            )
        };
    }
);
