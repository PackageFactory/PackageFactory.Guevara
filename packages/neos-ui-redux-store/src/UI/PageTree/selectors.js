import {$get} from 'plow-js';
import {createSelector} from 'reselect';

import {selectors as nodes} from '../../CR/Nodes/index';
import {getCurrentContentCanvasContextPath} from './../ContentCanvas/selectors';

export const getFocused = $get('ui.pageTree.isFocused');
export const getUncollapsed = $get('ui.pageTree.uncollapsed');
export const getLoading = $get('ui.pageTree.loading');
export const getErrors = $get('ui.pageTree.errors');

export const getFocusedNodeContextPathSelector = createSelector(
    [
        getFocused
    ],
    focusedNodeContextPath => focusedNodeContextPath
);

export const getUncollapsedContextPaths = createSelector(
    [
        getUncollapsed
    ],
    list => list.toJS()
);

export const getIsLoading = createSelector(
    [
        getLoading
    ],
    list => Boolean(list.toJS().length)
);

//
// TODO: NODETYPE REFACTORING - Fix calls of this
//
export const getTreeNodeSelector = createSelector(
    [
        getCurrentContentCanvasContextPath,
        getFocused,
        getUncollapsed,
        getLoading,
        getErrors,
        state => state
    ],
    (
        activeNodeContextPath,
        focusedNodeContextPath,
        uncollapsedNodeContextPaths,
        loadingNodeContextPaths,
        errorNodeContextPaths,
        state
    ) => (contextPath, nodeTypeFilter = []) => {
        //
        // Try to grab the node
        //
        const node = nodes.byContextPathSelector(contextPath)(state);
        const isNodeTypeValid = nodeType => Boolean(
            nodeTypeFilter.length === 0 ||
            nodeTypeFilter.indexOf(nodeType) > -1
        );

        //
        // Check if the requested node is existent and has the correct node type
        //
        if (node && isNodeTypeValid(node.nodeType)) {
            //
            // Check for valid child nodes
            //
            const validChildren = node.children.filter(node => isNodeTypeValid(node.nodeType));
            const {contextPath, label, uri} = node;

            //
            // Turn the node into a data structure, that can be consumed by a tree view
            //
            return {
                contextPath,
                label,
                uri,
                isActive: contextPath === activeNodeContextPath,
                isFocused: contextPath === focusedNodeContextPath,
                isCollapsed: !uncollapsedNodeContextPaths.contains(contextPath),
                isLoading: loadingNodeContextPaths.contains(contextPath),
                hasError: errorNodeContextPaths.contains(contextPath),
                hasChildren: validChildren.length > 0,
                children: validChildren.map(node => $get('contextPath', node))
            };
        }

        return null;
    }
);
