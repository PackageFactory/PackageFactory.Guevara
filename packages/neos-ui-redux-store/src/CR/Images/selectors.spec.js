import Immutable from 'immutable';
import {$all, $set} from 'plow-js';

import * as selectors from './selectors.js';

test(`should export selectors`, () => {
    expect(selectors).not.toBe(undefined);
    expect(typeof (selectors.imageLookup)).toBe('function');
});

test(`imageByUuid selector should return function that returns image data for a given uuid in the given state`, () => {
    const state = Immutable.fromJS($all(
        $set('cr.images.byUuid.5e798eb0-dc1c-40d6-b1f7-f6c046a7a8d7', {
            status: 'LOADED',
            some: 'value'
        }),
        {}
    ));
    const byExistentUuid = selectors.imageByUuid('5e798eb0-dc1c-40d6-b1f7-f6c046a7a8d7');
    const byNonExistentUuid = selectors.imageByUuid('355a1ee1-638a-4453-9cb5-e5735aa19493');

    expect(byExistentUuid(state)).toMatchSnapshot();
    expect(byNonExistentUuid(state)).toBe(undefined);
});

test(`imageLookup selector should return function that returns image data for the given state and uuid`, () => {
    const state = Immutable.fromJS($all(
        $set('cr.images.byUuid.5e798eb0-dc1c-40d6-b1f7-f6c046a7a8d7', {
            status: 'LOADED',
            some: 'value'
        }),
        {}
    ));
    const byUuid = selectors.imageLookup(state);

    expect(byUuid('5e798eb0-dc1c-40d6-b1f7-f6c046a7a8d7')).toMatchSnapshot();
    expect(byUuid('355a1ee1-638a-4453-9cb5-e5735aa19493')).toBe(undefined);
});
