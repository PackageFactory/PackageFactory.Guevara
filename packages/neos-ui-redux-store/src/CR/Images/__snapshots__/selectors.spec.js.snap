// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imageByUuid selector should return function that returns image data for a given uuid in the given state 1`] = `
Immutable.Map {
  "status": "LOADED",
  "some": "value",
}
`;

exports[`imageLookup selector should return function that returns image data for the given state and uuid 1`] = `
Immutable.Map {
  "status": "LOADED",
  "some": "value",
}
`;
